hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOperator.java

```diff
@@ -175,6 +183,16 @@ public EntryOperator operateOnKey(Data dataKey) {
         }
 
         oldValue = recordStore.get(dataKey, backup, callerAddress, false);
+
+        // Not required for OBJECT as #getOrNullOldValue() would return null in any case where mutation would be problematic
+        if (!readOnly && hasEventRegistration && inMemoryFormat != OBJECT) {
+            // This is a clone operation, but it's likely more performant to query the recordStore for the serialized value than
+            // regenerate it from oldValue
+            oldValueClone =  recordStore.get(dataKey, backup, callerAddress, false);
```

Comment:
As far as i see, `recordStore.get(dataKey, backup, callerAddress, false)` can also call user code via interceptor in it.
Instead i suggest to create a heap copy of data object with `mapServiceContext.toData(oldValue)`.
Since we eventually do it while publishing events, taking it here will make it perform similar with what we have currently i think.

Reply:
> As far as i see, `recordStore.get(dataKey, backup, callerAddress, false)` can also call user code via interceptor in it. Instead i suggest to create a heap copy of data object with `mapServiceContext.toData(oldValue)`. Since we eventually do it while publishing events, taking it here will make it perform similar with what we have currently i think.

@ahmetmircik Thanks - this works as well, but I'm not sure I quite follow what the difference is - are you able to explain, please?

Reply:
Sure, i see 3 main diffs: 
- `recordStore.get` calls interceptor in it. It is user-code and we shouldn't call it twice from the perspective
 of expected behavior. 
- Since interceptor is user-code, it can be slow for some reason.
- Eventually, when publishing events, we will do `mapServiceContext.toData(oldValue)` conversion. Bringing it to a previous line makes code perform similar while fixing the issue.

Reply:
> * `recordStore.get` calls interceptor in it. It is user-code and we shouldn't call it twice from the perspective of expected behavior.

Thanks! This makes sense, I've also added an extra test to cover this behaviour as well

592f18e7513809c272055d024e406f89b6cb7d3a